Hostap glue module

jukkar opened the pull request:
Add Kconfig + CMakeLists.txt files for hostap module. Allows compilation of hostap with zephyr but does not do anything useful yet. Integration code for wifi drivers is coming in subsequent PRs.

sr1dh48r committed:
hostap: Add build support for Zephyr

Only adds basic build support using Zephyr. Crypto support is disabled
till the MbedTLS integration is complete.

Signed-off-by: Sridhar Nuvusetty <sridhar.nuvusetty@nordicsemi.no>
Signed-off-by: Sachin Kulkarni <sachin.kulkarni@nordicsemi.no>
Signed-off-by: Ravi Dondaputi <ravi.dondaputi@nordicsemi.no>
Signed-off-by: Krishna T <krishna.t@nordicsemi.no>

jukkar committed:
manifest: Add hostap module

Add new hostap module that will support Wi-Fi supplicant
functionality.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

krish2718 commented on modules/hostap/CMakeLists.txt:
```diff
@@ -0,0 +1,90 @@
+	${COMMON_SRC_BASE}/utils/eloop.c
+	${COMMON_SRC_BASE}/utils/os_zephyr.c
+	${COMMON_SRC_BASE}/utils/wpa_debug_zephyr.c
+	${COMMON_SRC_BASE}/utils/radiotap.c
```
Recently, I had removed this, as its not needed.

jukkar replied:
Ok, will remove it then too.

rlubos commented on modules/hostap/Kconfig:
```diff
@@ -0,0 +1,64 @@
+	int "Stack size for wpa_supplicant thread"
+	default 8192
+
+config NET_SOCKETPAIR_BUFFER_SIZE
```
Since we're working directly in the Zephyr tree now - should all those defaults for Kconfigs defined in Zephyr be actually set where the individual entries are defined (default 4096 if WIFI_NM_WPA_SUPPLICANT for example)? This default redefinition works in NCS context, as NCS Kconfig files are processed before Zephyr (AFAIK Kconfig will apply the first default it finds), but is it guaranteed to work if we provide custom defaults in Zephyr itself? @tejlmand What's your say on this?

jukkar replied:
I temporarily moved various defaults to their respective Kconfig file. I think we need to revisit those values as I suspect we can tweak them more.
Also I am hoping to get rid of socketpair use between hostap and its handler thread. As they running in the same "process", we can use more lightweight IPC like k_fifo. This would mean that we could most probably use the default values already supplied in zephyr Kconfig files (guessing here) without any need to tweak those. Anyway, that is in future PRs to change.

jukkar committed:
hostap: Move the relevant config options away from hostap

Moving the Zephyr specific config options from
modules/hostap/Kconfig to corresponding Kconfig where the
option is specified.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

dleach02 commented on modules/hostap/Kconfig:
```diff
@@ -0,0 +1,50 @@
+
+# Supplicant API is stack heavy (buffers + snprintfs) and control interface
+# uses socketpair which pushes the stack usage causing overflow for 2048 bytes.
+# So we set SYSTEM_WORKQUEUE_STACK_SIZE default to 2560 in kernel/Kconfig
```
such a specific number 2560... is there any indication on how much leeway is available with this number?

jukkar replied:
@krish2718 do you know the history behind this value?

krish2718 replied:
2048 wasn't working, so, I tried few combinations and IIRC ~2300 should work as well, just to be safe, I added 2048 + 512 = 2560.

dleach02 commented on modules/hostap/CMakeLists.txt:
```diff
@@ -0,0 +1,89 @@
+#
+# Copyright (c) 2022 Nordic Semiconductor ASA
```
do you want to make this a 2023 copyright?

jukkar replied:
Yes, will do.

dleach02 commented on modules/hostap/Kconfig:
```diff
@@ -0,0 +1,50 @@
+# WPA Supplicant configuration options
+#
+# Copyright (c) 2022 Nordic Semiconductor
```
2023?

tejlmand commented on scripts/ci/check_compliance.py:
```diff
@@ -679,6 +682,11 @@ def check_no_undef_outside_kconfig(self, kconf):
         "MYFEATURE",
         "MY_DRIVER_0",
         "NORMAL_SLEEP",  # #defined by RV32M1 in ext/
+        "NO_CONFIG_BLOBS",      # Used in modules/hostap
+        "NO_CONFIG_WRITE",      # Used in modules/hostap
+        "NO_PBKDF2",            # Used in modules/hostap
+        "NO_RANDOM_POOL",       # Used in modules/hostap
+        "NO_WPA",               # Used in modules/hostap
```
> The upstream hostap project uses various CONFIG_* options
that we need to whitelist.

This is only a problem because you directly set those as compile definitions here:
https://github.com/zephyrproject-rtos/zephyr/blob/9062595a5be321c3b3c0c7ab05813253a3dc17d1/modules/hostap/CMakeLists.txt#L39

If instead the hostap Kconfig file defines equivalent Kconfig options as mentioned in other comment, then those changes / whitelisting wouldn't be needed.

krish2718 replied:
There are way too many configuration options provided by WPA suppplicant, so, IMHO, we shouldn't expose all of them as Kconfig options, instead abstract them with higher level defines.

cfriedt replied:
It would be good to leave the module design flexible so that Kconfig can be introduced on an as-needed basis. Likely the most common defines for the Zephyr platform can be selected from the start, and bring out only the most relevant / important ones initially. Not a huge worry from my perspective.

tejlmand replied:
> There are way too many configuration options provided by WPA suppplicant, so, IMHO, we shouldn't expose all of them as Kconfig options, instead abstract them with higher level defines.

Unsure what you mean with expose.
You can introduce them as promptless, meaning they will have the setting define, possibly with a help text describing it's functionality.
Being promptless means the value is not directly user configurable.

Minimal example without help text could be:
```
config NO_WPA
        bool
        default y
```

A big reason for having the hostap configs mapped into Kconfig is to expose to others that this Kconfig setting name is already in use.

Somewhere else in our Zephyr Kconfig code base someone could introduce a:
```
config CRYPTO_INTERNAL
    bool "FOO internal crypto implementation"
    default n
    help
      FOO internal crypto implementation having nothing in common with hostap, 
      but having name collision because it was not visible that same name is used by hostap. 
```

or someone may introduce a new Kconfig in Zephyr which are also used in hostap, but not exposed as such.

Hostap tests for `CONFIG_ECC`, `CONFIG_ECC`, `CONFIG_SHA384`, and similar.
Luckily Zephyr uses more specific names, such as `BT_MESH`, `BT_ECC`, and so.
But imagine Zephyr having a Kconfig being set to true which collides with a config that is used by hostap.

I did notice some of those are used in files we include in the Zephyr library.

krish2718 replied:
> Unsure what you mean with expose.

They don't have a corresponding Zephyr Kconfig.

> A big reason for having the hostap configs mapped into Kconfig is to expose to others that this Kconfig setting name is already in use.

Understood.

> I did notice some of those are used in files we include in the Zephyr library.

Oh, haven't realized that.

tejlmand reviewed (dismissed):
some minor observations / nits before a +1.

Generally good looking 🙂

tejlmand commented on modules/hostap/CMakeLists.txt:
```diff
@@ -0,0 +1,89 @@
+	CONFIG_NO_CONFIG_WRITE
+	CONFIG_NO_CONFIG_BLOBS
+	CONFIG_CTRL_IFACE
+	CONFIG_CTRL_IFACE_ZEPHYR
```
I don't see this define being used in the hostap source code being included in the zephyr library.

tejlmand commented on modules/hostap/CMakeLists.txt:
```diff
@@ -0,0 +1,89 @@
+	CONFIG_CTRL_IFACE_ZEPHYR
+	CONFIG_NO_RANDOM_POOL
+	CONFIG_NO_WPA
+	CONFIG_CRYPTO_INTERNAL
```
seems to be used only by sha1-internal.c, a file not included in the zephyr library.

tejlmand reviewed (commented):
seems there are a couple of defines which aare not used by the source code included in the Zephyr library.

Can reason be given ?

jukkar commented on modules/hostap/CMakeLists.txt:
```diff
@@ -0,0 +1,89 @@
+set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -DMISSING_SYSCALL_NAMES")
+
+zephyr_compile_definitions(
+	CONFIG_ZEPHYR
```
Yes, I agree, the name is a bit too generic. I suggest that we use this name for this PR, I can change it to something more appropriate in the next PR that changes the hostap repo (coming very soon).
